modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1233 @@
+  this._addressStorage = new sys_sockaddr_t(address);
+}
+
+/*
+  Returns a new record of type :type:`ipAddr` prvoided `host`, `port`
+  and `family`. this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.create("127.0.0.1", 8111, IPFamily.IPv4)
+
+  :arg host: address string in dot-dash or colon notation depending on family.
+  :type host: `string`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :arg family: value of IP Family
+  :type family: :type:`IPFamily`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.create(host: string = "127.0.0.1", port: uint(16) = 8000, family: IPFamily = IPFamily.IPv4): ipAddr throws {
+  var addressStorage = new sys_sockaddr_t();
+  addressStorage.set(host.c_str(), port, family:c_int);
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns a new record of type `ipAddr` prvoided `host` and `port`.
+  The family type is assumed based on `host` which is a standard address.
+  this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.ipv4(IPv4LocalHost, 8111)
+
+  :arg host: standard ipv4 address.
+  :type host: `ipv4Addr`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.ipv4(host: ipv4Addr, port: uint(16) = 8000): ipAddr throws {
+  var addressStorage = new sys_sockaddr_t();
+  addressStorage.set(host, port);
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns a new record of type `ipAddr` prvoided `host` and `port`.
+  The family type is assumed based on `host` which is a standard address.
+  this function is equivalent to the following code:
+
+  .. code-block:: Chapel
+
+    ipAddr.ipv6(IPv6LocalHost, 8111)
+
+  :arg host: standard ipv6 address.
+  :type host: `ipv6Addr`
+  :arg port: network address's port.
+  :type port: `uint(16)`
+  :return: address instance.
+  :rtype: `ipAddr`
+  :throws SystemError: Upon incompatible `host`, `port` or `family`
+*/
+proc type ipAddr.ipv6(host: ipv6Addr, port: uint(16) = 8000): ipAddr throws {
+  var addressStorage = new sys_sockaddr_t();
+  addressStorage.set(host, port);
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns the family type of address.
+  :return: family type of address
+  :rtype: :type:`IPFamily`
+*/
+proc ipAddr.family throws {
+  return _addressStorage.family:IPFamily;
+}
+
+/*
+  Returns the host address.
+  :return: host address
+  :rtype: `string`
+*/
+proc ipAddr.host throws {
+  return _addressStorage.numericHost();
+}
+
+/*
+  Returns the `port` stored in record.
+  :return: Returns numeric port.
+  :rtype: `uint(16)`
+*/
+proc ipAddr.port throws {
+  return _addressStorage.port();
+}
+
+/* compare ipAddr */
+inline operator !=(in lhs: ipAddr,in rhs: ipAddr) throws {
```

Comment:
Why does this throw?

Reply:
it's possible that the user might have not configured the ipAddr properly so the getters for `family`, `host` and `port` are throwing that's why the comparison operator is also throwing

Reply:
I can add in a `try!` here because the `ipAddr` will set properly but my thoughts were let's say a user does something like
```chapel
var addr1:ipAddr, addr2:ipAddr;
writeln(addr1 != addr2);
```
Basically, it will throw or halt (will it ?) if we use `try!` when the user uses an uninitialized `ipAddr` record.

Reply:
We can create a `ipAddr.init()` that default initializes the record with 0.0.0.0 e.g. so that these do not have to throw/halt in that case. IMO it will be better if the initializer guarantees that the ipAddr is valid enough for the getters not to throw, and then neither comparison nor comparison will throw.